---
title: "Geographic_distance_analysis"
author: "Jcb Nunez"
date: "12/30/2020"
output: html_document
---

# Interact for testing
```{sh, eval = F}
ijob -A jcbnunez -c 10 --mem=60G  --partition=standard
module load gcc/7.1.0  
module load openmpi/3.1.4
module load gdal
module load proj
module load htslib/1.9
module load bcftools/1.9
module load intel/18.0  
module load intelmpi/18.0
module load R/4.0.0
R
```


# Load packages
```{r}
library(data.table)
library(sp)
library(foreach)
library(tidyverse)
library(magrittr)
library(patchwork)
```

# Load Data
```{r}
test=GeoAnalysis
  
#User defined input
priv.dt <- fread("PoolSNP.goodSamps.0.001.delim") #file
ind.filter <- "good.samps" #good.samps, all.samps
caller <- "PoolSNP" #PoolSNPs, SNAPE
maf_thresh <- 0.05 #0.001, 0.05

pops <- names(fread("bcftools view -S ./good.samps.txt /project/berglandlab/DEST/vcf/dest.PoolSeq.PoolSNP.001.50.10Nov2020.ann.vcf.gz | head -n 40",
        nrows=1, skip="#CHR"))[-(1:9)]

```

#Parse data -- Add names
```{r}
if(caller=="PoolSNP") {
  setnames(priv.dt, names(priv.dt), paste("V", 2:9, sep=""))

priv.dt[,V8:=paste(V9, paste(V6, V7, V8, sep=""), sep=";")]

## Add names to the prov object
names(priv.dt) = c(
  "chr",
  "pos",
  "nPop",
  "missingPop",
  "ref",
  "alt",
  "afs_mut_pop",
  "afs_zero"
)} #close "if "-- PoolSNP 

if(caller=="SNAPE") {  
  setnames(priv.dt, names(priv.dt), paste("V", 2:9, sep=""))

priv.dt[,V8:=paste(V9, paste(V6, V7, V8, sep=""), sep=";")]

## Add names to the prov object
names(priv.dt) = c(
  "AF_bin_aob",
  "chr",
  "pos",
  "nPop",
  "missingPop",
  "ref",
  "alt",
  "afs_mut_pop",
  "afs_zero"
)} # close else if -- SNAPE

```

#Parse data -- Select Autosomes
```{r}
## Restrict any further analyses to the 4 main chromosomes				
priv.dt = priv.dt[which(priv.dt$chr %in% c("2L","2R","3L","3R")),]					
```

# Parse data - get distance
```{r}
### samps
  samps <- fread("/scratch/yey2sn/DEST/populationInfo/samps.csv")
  setkey(samps, sampleId)

### get distances
  pw.dist <- spDists(x=as.matrix(samps[set%in%c("DrosRTEC", "DrosEU"), c("long", "lat"), with=F]), longlat=T)
  diag(pw.dist) <- NA
  pw.dist[lower.tri(pw.dist)] <- NA
```


#Start Geographical Sampling
```{r}
priv.dt.small <- priv.dt[,list(pops=sample(afs_mut_pop, 100, replace=T), n=.N), list(chr=chr, nPop=nPop)]
setkey(priv.dt.small, chr)

priv.dt.small <- na.omit(priv.dt.small)
dim(priv.dt.small)[1]
```

# Generate Function
```{r}
  getmean <- function(x) {
    # x <- priv.dt.small.tmp$pops
    af_string <- tstrsplit(x, ";")[[1]]
    mean(as.numeric(unlist(tstrsplit(af_string, "\\+")[-1])))
  }
```

# Apply function
```{r}
  priv.dt.small[,id:=c(1:dim(priv.dt.small)[1])]

o <- foreach(i=priv.dt.small$id)%dopar%{
    #i<-201
    if(i%%2==0) message(paste(i, dim(priv.dt.small)[1], sep=" / "))

    ### get average allele frequency
      priv.dt.small.tmp <- priv.dt.small[id==i]
      priv.dt.small.tmp[,af:=getmean(pops)]

    ### obs
      set.obs <- samps[J(pops[as.numeric(unlist(tstrsplit(priv.dt.small.tmp$pops, ";")[-c(1,2)]))])]

      pw.dist.obs <- spDists(x=as.matrix(set.obs[,c("long", "lat"), with=F]), longlat=T)

    ### exp
      set.exp <- samps[J(sample(pops, nrow(set.obs)))]

      pw.dist.exp <- spDists(x=as.matrix(set.exp[,c("long", "lat"), with=F]), longlat=T)

      o <- cbind(rbind(priv.dt.small.tmp, data.table(chr=priv.dt.small.tmp$chr, nPop=priv.dt.small.tmp$nPop, pops=NA, n=NA, id=NA), fill=T),
          data.table(set=c("obs", "exp"),
                      meanDist=c(mean(pw.dist.obs[lower.tri(pw.dist.obs)]),
                                mean(pw.dist.exp[lower.tri(pw.dist.exp)])),

                      minDist=c(min(pw.dist.obs[lower.tri(pw.dist.obs)]),
                                min(pw.dist.exp[lower.tri(pw.dist.exp)])),

                      maxDist=c(max(pw.dist.obs[lower.tri(pw.dist.obs)]),
                                  max(pw.dist.exp[lower.tri(pw.dist.exp)]))#,

                      #cc_equal=c(length(unique(set.obs$Continental_clusters))==1,
                      #          length(unique(set.exp$Continental_clusters))==1)
                     ))
      o[,mt:=tstrsplit(pops, ";")[[2]]]
      o[,caller:=caller]
      o[,maf:=maf_thresh]
      return(o)
}

o <- rbindlist(o)

save(o, file= paste(test,
                    caller, 
                    maf_thresh, 
                    ind.filter, 
                    "Rdata", 
                    sep = "."))

```

# Test Plot
```{r}
##o %>%
##  group_by(nPop,set, ) %>%
##  summarise(dist_m = mean(meanDist)) %>%
##  ggplot(aes(x=nPop,
##             y=dist_m,
##            # ymin= dist_l,
##            # ymax = dist_h,
##             fill=set,
##             color= set)) +
##  geom_line() -> test.1
##
##ggsave("test.1.png", test.1, height = 3, width = 6)
##
##
##o %>%
##  mutate(AF_fold = ifelse(.$af > 0.5, (1-.$af), .$af)) %>% 
##  group_by(AF_fold,set, ) %>%
##  summarise(dist_m = mean(meanDist)) %>%
##  ggplot(aes(x=AF_fold,
##             y=dist_m,
##            # ymin= dist_l,
##            # ymax = dist_h,
##             fill=set,
##             color= set)) +
##  geom_smooth() -> test.2
##
##ggsave("test.2.png", test.2, height = 3, width = 6)
##
##
```

